crates/astria-composer/src/searcher/executor.rs

```diff
@@ -0,0 +1,296 @@
+    pub(super) fn subscribe(&self) -> watch::Receiver<Status> {
+        self.status.subscribe()
+    }
+
+    /// Gets the next nonce to sign over if it exists and increments the stored nonce counter
+    fn get_and_increment_nonce(&mut self) -> Option<u32> {
+        self.nonce.map(|curr_nonce| {
+            self.nonce = Some(curr_nonce + 1);
+            curr_nonce
+        })
+    }
+
+    /// Sugmits a signed transaction to the sequencer node.
+    /// TODO: handle failed tx submission due to nonce
+    async fn submit_tx(&self, signed_tx: SignedTransaction) -> eyre::Result<()> {
+        let rsp = self
+            .sequencer_client
+            .submit_transaction_sync(signed_tx)
+            .await
+            .wrap_err("failed submitting transaction to sequencer")?;
+        if rsp.code.is_err() {
+            bail!(
+                "submitting transaction to sequencer returned with error code; code: `{code}`; \
+                 log: `{log}`; hash: `{hash}`",
+                code = rsp.code.value(),
+                log = rsp.log,
+                hash = rsp.hash,
+            );
+        }
+        Ok(())
+    }
+
+    /// Signs and submits the bundle of actions to the sequencer.
+    async fn sign_and_submit(&mut self, actions: Vec<Action>) -> eyre::Result<()> {
+        let nonce = self
+            .get_and_increment_nonce()
+            .ok_or(eyre!("no nonce stored; cannot process bundle"))?;
+
+        let tx = UnsignedTransaction {
+            nonce,
+            actions,
+        }
+        .into_signed(&self.sequencer_key);
+
+        self.submit_tx(tx)
+            .await
+            .wrap_err("failed submitting signed actions to sequencer")?;
+        Ok(())
+    }
+
+    /// Run the Executor loop, calling `process_bundle` on each bundle received from the channel.
+    ///
+    /// # Errors
+    /// An error is returned if connecting to the sequencer fails.
+    pub(super) async fn run_until_stopped(mut self) -> eyre::Result<()> {
+        // set up connection to sequencer
+        self.init_nonce_from_sequencer(5, Duration::from_secs(5), 2.0)
+            .await
+            .wrap_err("failed retrieving initial nonce from sequencer")?;
+
+        while let Some(bundle) = self.executor_rx.recv().await {
+            if let Err(e) = self.sign_and_submit(bundle).await {
+                // FIXME: currently this will fail both when there is an issue with the nonce and
+                // when unable to reach the sequencer. As there is currently no error returned by
+                // the sequencer for invalid nonces, there is nothing to handle. This should be
+                // changed after #364 is merged in a followup PR to handle nonce failues.
+                error!(
+                    error.message = %e,
+                    error.cause_chain = ?e,
+                    "failed submitting the bundle to the sequencer; bailing",
+                );
+                break;
```

Comment:
Note that I added this. Upon further inspection some of the errors are critical,  preventing the service from going on.